File: src/term.rs
use std::os::fd::RawFd;
use std::sync::atomic::{AtomicBool, Ordering};

pub const KEY_NONE: i32 = -1;
pub const KEY_ESC: i32 = 0x1b;
pub const KEY_UP: i32 = 0x1001;
pub const KEY_DOWN: i32 = 0x1002;
pub const KEY_LEFT: i32 = 0x1003;
pub const KEY_RIGHT: i32 = 0x1004;
pub const KEY_ENTER: i32 = 0x1005;
pub const KEY_BACKSPACE: i32 = 0x1006;
pub const KEY_CHAR: i32 = 0x1007;

static mut SAVED: std::mem::MaybeUninit<libc::termios> = std::mem::MaybeUninit::uninit();
static HAVE_SAVED: AtomicBool = AtomicBool::new(false);

pub fn term_winsize(fd: RawFd, rows: &mut u32, cols: &mut u32) -> bool {
    let mut ws: libc::winsize = unsafe { std::mem::zeroed() };
    if unsafe { libc::ioctl(fd, libc::TIOCGWINSZ, &mut ws) } != 0 || ws.ws_row == 0 || ws.ws_col == 0 {
        return false;
    }
    *rows = ws.ws_row as u32;
    *cols = ws.ws_col as u32;
    true
}

/// Terminal pixel dimensions (total) from TIOCGWINSZ, if the terminal
/// reports them. Returns (0, 0) when unavailable.
pub fn term_winsize_px(fd: RawFd) -> (u32, u32) {
    let mut ws: libc::winsize = unsafe { std::mem::zeroed() };
    if unsafe { libc::ioctl(fd, libc::TIOCGWINSZ, &mut ws) } != 0 {
        return (0, 0);
    }
    (ws.ws_xpixel as u32, ws.ws_ypixel as u32)
}

pub fn term_raw_enter(fd: RawFd) -> bool {
    let mut t: libc::termios = unsafe { std::mem::zeroed() };
    if unsafe { libc::tcgetattr(fd, &mut t) } != 0 {
        return false;
    }
    unsafe {
        std::ptr::addr_of_mut!(SAVED).cast::<libc::termios>().write(t);
    }
    HAVE_SAVED.store(true, Ordering::Release);
    t.c_lflag &= !(libc::ICANON | libc::ECHO | libc::ISIG);
    t.c_iflag &= !(libc::IXON | libc::ICRNL);
    t.c_cc[libc::VMIN] = 0;
    t.c_cc[libc::VTIME] = 0;
    unsafe { libc::tcsetattr(fd, libc::TCSANOW, &t) == 0 }
}

pub fn term_raw_restore(fd: RawFd) {
    if HAVE_SAVED.load(Ordering::Acquire) {
        unsafe {
            let p = std::ptr::addr_of!(SAVED).cast::<libc::termios>();
            libc::tcsetattr(fd, libc::TCSANOW, p);
        }
    }
}

fn poll_readable(fd: RawFd, timeout_ms: i32) -> bool {
    let mut pfd = libc::pollfd {
        fd,
        events: libc::POLLIN,
        revents: 0,
    };
    unsafe { libc::poll(&mut pfd, 1, timeout_ms) > 0 }
}

fn read_byte(fd: RawFd) -> Option<u8> {
    let mut c: u8 = 0;
    if unsafe { libc::read(fd, &mut c as *mut u8 as *mut libc::c_void, 1) } == 1 {
        Some(c)
    } else {
        None
    }
}

#[allow(dead_code)]
pub fn term_read_key(fd: RawFd) -> i32 {
    let c = match read_byte(fd) {
        Some(c) => c,
        None => return KEY_NONE,
    };
    if c != 0x1b {
        return c as i32;
    }
    if poll_readable(fd, 30) {
        if let Some(c2) = read_byte(fd) {
            if c2 == b'[' && poll_readable(fd, 30) {
                if let Some(c3) = read_byte(fd) {
                    return match c3 {
                        b'A' => KEY_UP,
                        b'B' => KEY_DOWN,
                        b'C' => KEY_RIGHT,
                        b'D' => KEY_LEFT,
                        _ => c3 as i32,
                    };
                }
            }
        }
    }
    KEY_ESC
}

/// Reads a key press. Returns `(key, codepoint bytes, length of codepoint)`.
/// When `key == KEY_CHAR`, `cp[..len]` holds the UTF-8 sequence.
pub fn term_read_codepoint(fd: RawFd, out: &mut [u8; 8]) -> (i32, usize) {
    let c = match read_byte(fd) {
        Some(c) => c,
        None => return (KEY_NONE, 0),
    };
    if c == 0x1b {
        if poll_readable(fd, 30) {
            if let Some(c2) = read_byte(fd) {
                if c2 == b'[' && poll_readable(fd, 30) {
                    if let Some(c3) = read_byte(fd) {
                        return match c3 {
                            b'A' => (KEY_UP, 0),
                            b'B' => (KEY_DOWN, 0),
                            b'C' => (KEY_RIGHT, 0),
                            b'D' => (KEY_LEFT, 0),
                            _ => (KEY_ESC, 0),
                        };
                    }
                }
            }
        }
        return (KEY_ESC, 0);
    }
    if c == 0x0d || c == 0x0a {
        return (KEY_ENTER, 0);
    }
    if c == 0x7f || c == 0x08 {
        return (KEY_BACKSPACE, 0);
    }
    if c < 0x20 {
        return (c as i32, 0);
    }

    let need = if c < 0x80 {
        1
    } else if (c & 0xE0) == 0xC0 {
        2
    } else if (c & 0xF0) == 0xE0 {
        3
    } else if (c & 0xF8) == 0xF0 {
        4
    } else {
        return (KEY_NONE, 0);
    };

    let mut got = 1usize;
    out[0] = c;
    while got < need {
        if !poll_readable(fd, 50) {
            break;
        }
        match read_byte(fd) {
            Some(cc) => {
                out[got] = cc;
                got += 1;
            }
            None => break,
        }
    }
    (KEY_CHAR, got)
}

File: src/term.c
#include <poll.h>
#include <string.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

#include "term.h"

static struct termios saved_term;
static int have_saved = 0;

int term_winsize(int fd, unsigned *rows, unsigned *cols) {
    struct winsize ws;
    memset(&ws, 0, sizeof ws);
    if (ioctl(fd, TIOCGWINSZ, &ws) != 0)
        return 0;
    if (ws.ws_row == 0 || ws.ws_col == 0)
        return 0;
    *rows = ws.ws_row;
    *cols = ws.ws_col;
    return 1;
}

double term_cell_aspect(int fd) {
    struct winsize ws;
    memset(&ws, 0, sizeof ws);
    if (ioctl(fd, TIOCGWINSZ, &ws) != 0)
        return 2.0;
    if (ws.ws_row == 0 || ws.ws_col == 0 || ws.ws_xpixel == 0 || ws.ws_ypixel == 0)
        return 2.0;
    double cw = (double)ws.ws_xpixel / (double)ws.ws_col;
    double ch = (double)ws.ws_ypixel / (double)ws.ws_row;
    if (!(cw > 0.0) || !(ch > 0.0))
        return 2.0;
    double r = ch / cw;
    if (r < 0.5)
        r = 0.5;
    if (r > 3.0)
        r = 3.0;
    return r;
}

int term_raw_enter(int fd) {
    struct termios t;
    if (tcgetattr(fd, &t) != 0)
        return 0;
    saved_term = t;
    have_saved = 1;
    t.c_lflag &= (unsigned)(~(ICANON | ECHO | ISIG));
    t.c_iflag &= (unsigned)(~(IXON | ICRNL));
    t.c_cc[VMIN] = 0;
    t.c_cc[VTIME] = 0;
    return tcsetattr(fd, TCSANOW, &t) == 0;
}

void term_raw_restore(int fd) {
    if (have_saved)
        tcsetattr(fd, TCSANOW, &saved_term);
}

static int poll_readable(int fd, int timeout_ms) {
    struct pollfd pfd;
    pfd.fd = fd;
    pfd.events = POLLIN;
    pfd.revents = 0;
    return poll(&pfd, 1, timeout_ms) > 0;
}

static int read_byte(int fd, uint8_t *out) {
    uint8_t c;
    if (read(fd, &c, 1) == 1) {
        *out = c;
        return 1;
    }
    return 0;
}

int term_read_key(int fd) {
    uint8_t c;
    if (!read_byte(fd, &c))
        return KEY_NONE;
    if (c != 0x1b)
        return (int)c;
    if (poll_readable(fd, 30)) {
        uint8_t c2;
        if (read_byte(fd, &c2)) {
            if (c2 == '[' && poll_readable(fd, 30)) {
                uint8_t c3;
                if (read_byte(fd, &c3)) {
                    switch (c3) {
                    case 'A':
                        return KEY_UP;
                    case 'B':
                        return KEY_DOWN;
                    case 'C':
                        return KEY_RIGHT;
                    case 'D':
                        return KEY_LEFT;
                    default:
                        return (int)c3;
                    }
                }
            }
        }
    }
    return KEY_ESC;
}

int term_read_codepoint(int fd, uint8_t out8[8], size_t *n) {
    uint8_t c;
    if (!read_byte(fd, &c)) {
        *n = 0;
        return KEY_NONE;
    }
    if (c == 0x1b) {
        if (poll_readable(fd, 30)) {
            uint8_t c2;
            if (read_byte(fd, &c2)) {
                if (c2 == '[' && poll_readable(fd, 30)) {
                    uint8_t c3;
                    if (read_byte(fd, &c3)) {
                        *n = 0;
                        switch (c3) {
                        case 'A':
                            return KEY_UP;
                        case 'B':
                            return KEY_DOWN;
                        case 'C':
                            return KEY_RIGHT;
                        case 'D':
                            return KEY_LEFT;
                        default:
                            return KEY_ESC;
                        }
                    }
                }
            }
        }
        *n = 0;
        return KEY_ESC;
    }
    if (c == 0x0d || c == 0x0a) {
        *n = 0;
        return KEY_ENTER;
    }
    if (c == 0x7f || c == 0x08) {
        *n = 0;
        return KEY_BACKSPACE;
    }
    if (c < 0x20) {
        *n = 0;
        return (int)c;
    }
    size_t need;
    if (c < 0x80)
        need = 1;
    else if ((c & 0xE0) == 0xC0)
        need = 2;
    else if ((c & 0xF0) == 0xE0)
        need = 3;
    else if ((c & 0xF8) == 0xF0)
        need = 4;
    else {
        *n = 0;
        return KEY_NONE;
    }
    out8[0] = c;
    size_t got = 1;
    while (got < need) {
        if (!poll_readable(fd, 50))
            break;
        uint8_t cc;
        if (!read_byte(fd, &cc))
            break;
        out8[got++] = cc;
    }
    *n = got;
    return KEY_CHAR;
}
